fbpic/particles/gathering/cuda_methods.py

```diff
@@ -335,3 +505,138 @@ def gather_field_gpu_cubic(x, y, z,
+
+    invdz, invdr : float (in meters^-1)
+        Inverse of the grid step along the considered direction
+
+    zmin, rmin : float (in meters)
+        Position of the edge of the simulation box along the
+        direction considered
+
+    Nz, Nr : int
+        Number of gridpoints along the considered direction
+
+    a_tuple, grad_a_r_tuple, grad_a_t_tuple, grad_a_z_tuple :
+        tuple of 2darray of complexs
+        The relevant fields on the interpolation grid for all the different modes
+
+    m: tuple
+        Indices of the azimuthal mode
+
+    a, grad_a_x, grad_a_y, grad_a_z : 1darray of floats
+        The relevant fields acting on the particles
+        (is modified by this function)
+
+    averaging : boolean, optional
+        Whether to average the new values with the old ones or to
+        discard the old values.
+    """
+
+
+    # Get the 1D CUDA grid
+    i = cuda.grid(1)
+    # Deposit the field per cell in parallel
+    # (for threads < number of particles)
+    if i < x.shape[0]:
+        # Preliminary arrays for the cylindrical conversion
+        # --------------------------------------------
+        # Position
+        xj = x[i]
+        yj = y[i]
+        zj = z[i]
+
+        # Cylindrical conversion
+        rj = math.sqrt(xj**2 + yj**2)
+        if (rj != 0.):
+            invr = 1./rj
+            cos = xj*invr  # Cosine
+            sin = yj*invr  # Sine
+        else:
+            cos = 1.
+            sin = 0.
+
+        # Get weights for the deposition
+        # --------------------------------------------
+        # Positions of the particle, in the cell unit
+        r_cell = invdr*(rj - rmin) - 0.5
+        z_cell = invdz*(zj - zmin) - 0.5
+
+        # Calculate the shape factors
+        Sr = cuda.local.array((4,), dtype=float64)
+        ir_lowest = int64(math.floor(r_cell)) - 1
+        r_local = r_cell-ir_lowest
+        Sr[0] = -1./6. * (r_local-2.)**3
+        Sr[1] = 1./6. * (3.*(r_local-1.)**3 - 6.*(r_local-1.)**2 + 4.)
+        Sr[2] = 1./6. * (3.*(2.-r_local)**3 - 6.*(2.-r_local)**2 + 4.)
+        Sr[3] = -1./6. * (1.-r_local)**3
+        Sz = cuda.local.array((4,), dtype=float64)
+        iz_lowest = int64(math.floor(z_cell)) - 1
+        z_local = z_cell-iz_lowest
+        Sz[0] = -1./6. * (z_local-2.)**3
+        Sz[1] = 1./6. * (3.*(z_local-1.)**3 - 6.*(z_local-1.)**2 + 4.)
+        Sz[2] = 1./6. * (3.*(2.-z_local)**3 - 6.*(2.-z_local)**2 + 4.)
+        Sz[3] = -1./6. * (1.-z_local)**3
+
+        # Envelope fields
+        # -------
+        F = 0.
+        Fr = 0.
+        Ft = 0.
+        Fz = 0.
+        for m in m_tuple:
+            # Add contribution from mode m
+            a_m = a_tuple[m]
+            grad_a_r_m = grad_a_r_tuple[m]
+            grad_a_t_m = grad_a_t_tuple[m]
+            grad_a_z_m = grad_a_z_tuple[m]
+            # Calculate azimuthal complex factor
+            exptheta_m = 1.
+            for _ in range(m):
+                exptheta_m *= (cos - 1.j*sin)
```

Comment:
Same comment: this could be replaced by
```
for _ in range(abs(m)):
    exptheta_m *= (cos - 1.j*sin)
if m < 0:
   exptheta_m = exp_theta.conjugate()
```